text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+
+## `core::future` module
+
+With all of the above task infrastructure in place, defining `Future` is
+straightforward.  Like `Iterator`, the `Future` trait has a single required
+method, `poll`, and a large number of provided methods (called "adapters" or
+"combinators"). We'll look first at `poll`:
+
+```rust
+pub trait Future {
+    /// The type of value produced on completion.
+    type Item;
+
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - `Poll::Pending` if the future is not ready yet.
+    /// - `Poll::Ready(val)` with the result `val` of this future if it completed.
+    ///
+    /// Once a future has completed, clients should not `poll` it again.
+    ///
+    /// When a future is not ready yet, `poll` returns `Poll::Pending`.
+    /// The future will *also* register the interest of the current task in the
+    /// value being produced. For example, if the future represents the availability
+    /// of data on a socket, then the task is recorded so that when data arrives,
+    /// it is woken up (via `cx.waker()`). Once a task has been woken up,
+    /// it should attempt to `poll` the future again, which may or may not
+    /// produce a final value at that time.
+    ///
+    /// Note that if `Pending` is returned it only means that the *current* task
+    /// (represented by the argument `cx`) will receive a notification. Tasks
+    /// from previous calls to `poll` will *not* receive notifications.
+    ///
+    /// # Runtime characteristics
+    ///
+    /// Futures alone are *inert*; they must be *actively* `poll`ed to make
+    /// progress, meaning that each time the current task is woken up, it should
+    /// actively re-`poll` pending futures that it still has an interest in.
+    /// Usually this is handled automatically by `async`/`await` notation or
+    /// via adapter methods. Executors ensure that each task is `poll`ed every
+    /// time a future internal to that task is ready to make progress.
+    ///
+    /// The `poll` function is not called repeatedly in a tight loop for
+    /// futures, but only whenever the future itself is ready, as signaled via
+    /// `cx.waker()`. If you're familiar with the `poll(2)` or `select(2)`
+    /// syscalls on Unix it's worth noting that futures typically do *not*
+    /// suffer the same problems of "all wakeups must poll all events"; they
+    /// are more like `epoll(4)`.
+    ///
+    /// An implementation of `poll` should strive to return quickly, and should
+    /// *never* block. Returning quickly prevents unnecessarily clogging up
+    /// threads or event loops. If it is known ahead of time that a call to
+    /// `poll` may end up taking awhile, the work should be offloaded to a
+    /// thread pool (or something similar) to ensure that `poll` can return
+    /// quickly.
```

Comment:
You may or may not want to clarify that you need a Future-aware thread pool. Most thread pool libraries currently available on crates.io either do not provide a way to synchronize with the end of the computation or do so by implicitly blocking, neither of which is appropriate here. Only some thread pools will be usable here.

Reply:
It should be possible to offload blocking work to any kind of thread pool, and "synchronizing" the result back to a future, using something like futures current `oneshot` channel:

You send the arguments of your computation, and a `oneshot::Sender<ReturnType>`, to the other thread. Once computation is completed, you'd call `tx.send(ret)`. In your `Future`, you'd just await the result from the `oneshot::Receiver`.

Reply:
You're right, this can be done with any kind of fire-and-forget API (and I think this is how current futures-aware thread pools are implemented). It may not be immediately obvious, however, and will not work with blocking APIs (scoped_threadpool, rayon...).

Again, I'm not 100% sure if clarifying this part is needed, just thought it would be another possible area for design documentation improvements.